File: bench_xlang/bench_cost_model.py
"""Cross-language cost-model audit: skip-flag pricing and Möbius
per-matrix costs.

Part 1 (deterministic, no timing): calls the Bulger-vs-Möbius selector
across a (r, K, N) grid with the self-matrix skip flags off (a first
pair: full triple) and on (a later broadcast pair: cross term only),
recording the chosen route and both predicted costs. Rows where the
choice flips are the cells the skip flags exist for.

Part 2 (timing): times ``method='mobius'`` on absolute densities under
three memo regimes -- ``cosine`` fresh (three matrices), ``oneSidedDenom``
fresh (cross + one self), ``cosine`` warm selves (cross only) -- giving
measured per-matrix costs to hold against the pricing's
``n_matrices / 3`` scaling of the fitted whole-triple constants.

Part 3 (timing): at each flip cell from part 1, times both forced
routes in the warm-selves regime and reports whether the flags-on
choice is the faster -- the behavioural claim the skip flags make.

Writes ``bench_cost_model_python.csv``. The MATLAB counterpart is
``bench_cost_model.m``.
"""
import csv
import os
import sys

import numpy as np

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "python"))
import mpt                                                  # noqa: E402
mpt.set_default(show_hints=False)   # dispatch prints would sit inside the timed closures
import warnings                                       # noqa: E402
from mpt._defaults import TruncationDefaultWarning    # noqa: E402
warnings.filterwarnings("ignore", category=TruncationDefaultWarning)
from mpt.tensor import _select_ma_inner_product_method      # noqa: E402
from bench_xlang import adaptive_time                       # noqa: E402


def build_abs(K, N, r, seed_shift=0):
    """Deterministic absolute density: K values per event, N events."""
    j = np.arange(K * N, dtype=float) + 1 + seed_shift
    vals = (2000.0 * np.mod(11.0 * j * j + 5.0 * j, 397.0) / 397.0
            ).reshape(K, N)
    return mpt.build_exp_tens([vals], None, [30.0], [r], [False],
                              [False], [0.0], verbose=False)


def selector(r, K_x, N_x, skip, K_y=None, N_y=None):
    if K_y is None:
        K_y = K_x
    if N_y is None:
        N_y = N_x
    chosen, pw, orbit = _select_ma_inner_product_method(
        r_vec=np.array([r]), k_vec=np.array([K_x]), A=1,
        N_x=N_x, N_y=N_y,
        any_per=False, any_rel_nonper=False, any_rel_per=False,
        sigma_over_P_max=0.0, user_method="auto",
        rel_vec=np.array([False]), k_vec_y=np.array([K_y]),
        return_costs=True,
        pw_skip_xx=skip, pw_skip_yy=skip,
        orbit_skip_xx=skip, orbit_skip_yy=skip,
    )
    return chosen, pw, orbit


def clear_memos(*densities):
    for d in densities:
        d._self_ip_cache.clear()
        if d._pruned_cached is not None and d._pruned_cached is not d:
            d._pruned_cached._self_ip_cache.clear()


def main():
    out_path = os.path.join(os.path.dirname(__file__),
                            "bench_cost_model_python.csv")
    rows = []
    flips = []

    # ---- Part 1: selector audit -----------------------------------
    print("== selector skip-flag audit (abs, A=1, N_x=N_y) ==")
    for r in (2, 3):
        for N in (4, 16):
            for K in (4, 6, 8, 12, 16, 24, 40):
                c_off, pw_off, orb_off = selector(r, K, N, skip=False)
                c_on, pw_on, orb_on = selector(r, K, N, skip=True)
                flip = c_off != c_on
                if flip:
                    flips.append((r, K, N, c_off, c_on))
                rows.append(dict(
                    language="python", part="selector", r=r, K=K, N=N,
                    regime="", chosen_off=c_off, chosen_on=c_on,
                    pw_off=f"{pw_off:.4g}", orbit_off=f"{orb_off:.4g}",
                    pw_on=f"{pw_on:.4g}", orbit_on=f"{orb_on:.4g}",
                    ms=""))
                mark = "  <-- flip" if flip else ""
                print(f"  r={r} K={K:3d} N={N:3d}: off={c_off:6s} "
                      f"on={c_on:6s}{mark}")

    # Asymmetric cells: the broadcast regime. A large shared X against
    # a small fixed query (K_y = 4, N_y = 3): the shared self matrix
    # dominates the full-triple Bulger price, so skipping it (a later
    # broadcast pair) is where the routing genuinely moves.
    print("\n== selector audit, asymmetric (K_y=4, N_y=3) ==")
    for r in (2, 3):
        for N_x in (8, 32):
            for K_x in (6, 8, 12, 16, 24, 40, 64):
                c_off, pw_off, orb_off = selector(r, K_x, N_x, False,
                                                  K_y=4, N_y=3)
                c_on, pw_on, orb_on = selector(r, K_x, N_x, True,
                                               K_y=4, N_y=3)
                flip = c_off != c_on
                if flip:
                    flips.append((r, K_x, N_x, c_off, c_on))
                rows.append(dict(
                    language="python", part="selector_asym", r=r,
                    K=K_x, N=N_x, regime="Ky4_Ny3",
                    chosen_off=c_off, chosen_on=c_on,
                    pw_off=f"{pw_off:.4g}", orbit_off=f"{orb_off:.4g}",
                    pw_on=f"{pw_on:.4g}", orbit_on=f"{orb_on:.4g}",
                    ms=""))
                mark = "  <-- flip" if flip else ""
                print(f"  r={r} Kx={K_x:3d} Nx={N_x:3d}: off={c_off:6s} "
                      f"on={c_on:6s}{mark}")

    # ---- Part 2: Möbius per-matrix timings ------------------------
    print("\n== Möbius per-matrix timings (abs, A=1, K=8) ==")
    for r in (2, 3):
        for N in (20, 60):
            d_x = build_abs(8, N, r)
            d_y = build_abs(8, N, r, seed_shift=137)

            def t_cos_fresh():
                clear_memos(d_x, d_y)
                return mpt.cos_sim_exp_tens(d_x, d_y, method="mobius",
                                            verbose=False)

            def t_osd_fresh():
                clear_memos(d_x, d_y)
                return mpt.cos_sim_exp_tens(
                    d_x, d_y, method="mobius",
                    normalize="oneSidedDenom", verbose=False)

            def t_cos_warm():
                # Memo persists across calls on the same objects: after
                # the first call, both self terms are cached and only
                # the cross matrix is computed.
                return mpt.cos_sim_exp_tens(d_x, d_y, method="mobius",
                                            verbose=False)

            regimes = (("cos_fresh_3mat", t_cos_fresh),
                       ("osd_fresh_2mat", t_osd_fresh),
                       ("cos_warm_1mat", t_cos_warm))
            times = {}
            for name, fn in regimes:
                t, _, _ = adaptive_time(fn)
                times[name] = t * 1000.0
                rows.append(dict(
                    language="python", part="mobius_matrices", r=r,
                    K=8, N=N, regime=name, chosen_off="", chosen_on="",
                    pw_off="", orbit_off="", pw_on="", orbit_on="",
                    ms=f"{t * 1000.0:.5f}"))
            t3, t2, t1 = (times["cos_fresh_3mat"],
                          times["osd_fresh_2mat"],
                          times["cos_warm_1mat"])
            print(f"  r={r} N={N:3d}: 3mat {t3:.3f}  2mat {t2:.3f}  "
                  f"1mat {t1:.3f} ms  ->  xy~{t1:.3f}  "
                  f"yy~{max(t2 - t1, 0):.3f}  xx~{max(t3 - t2, 0):.3f}")

    # ---- Part 3: flip-cell behavioural check ----------------------
    print("\n== flip-cell forced-route timings (warm selves) ==")
    for (r, K, N, c_off, c_on) in flips[:4]:
        d_x = build_abs(K, N, r)
        d_y = build_abs(4, 3, r, seed_shift=137)
        forced = {}
        for meth in ("bulger", "mobius"):
            mpt.cos_sim_exp_tens(d_x, d_y, method=meth, verbose=False)

            def fn(meth=meth):
                return mpt.cos_sim_exp_tens(d_x, d_y, method=meth,
                                            verbose=False)
            t, _, _ = adaptive_time(fn)
            forced[meth] = t * 1000.0
            rows.append(dict(
                language="python", part="flip_check", r=r, K=K, N=N,
                regime=f"forced_{meth}_warm", chosen_off=c_off,
                chosen_on=c_on, pw_off="", orbit_off="", pw_on="",
                orbit_on="", ms=f"{t * 1000.0:.5f}"))
        faster = min(forced, key=forced.get)
        ok = "OK" if faster == c_on else "MISPICK"
        print(f"  r={r} K={K} N={N}: off->{c_off}, on->{c_on}; "
              f"warm bulger {forced['bulger']:.3f} ms, "
              f"mobius {forced['mobius']:.3f} ms; faster={faster} [{ok}]")

    with open(out_path, "w", newline="") as fh:
        writer = csv.DictWriter(fh, fieldnames=list(rows[0].keys()))
        writer.writeheader()
        writer.writerows(rows)
    print(f"\nwrote {out_path}")


if __name__ == "__main__":
    main()

File: bench_xlang/bench_cost_model.m
%% bench_cost_model.m — skip-flag pricing audit and Möbius per-matrix costs
%
%  MATLAB counterpart of bench_cost_model.py. Three parts:
%
%  1. Selector audit (deterministic): internal.selectMaInnerProductMethod
%     across a (r, K, N) grid with the self-matrix skip flags off (a
%     first pair: full triple) and on (a later broadcast pair: cross
%     only); rows where the routing flips are the cells the flags exist
%     for.
%  2. Möbius per-matrix timings: method = 'mobius' on absolute
%     densities under cosine-fresh (three matrices), oneSidedDenom-fresh
%     (cross + one self), and cosine-warm-selves (cross only; the memo
%     threaded via the cache-carrying outputs), giving measured
%     per-matrix costs to hold against the pricing's nMatrices/3
%     scaling of the fitted whole-triple constants.
%  3. Flip-cell behavioural check: both forced routes timed in the
%     warm-selves regime; the flags-on choice should be the faster.
%
%  Writes bench_cost_model_matlab.csv. Requires MATLAB
%  (internal.timeRepeated and +mobius use arguments blocks; not
%  Octave-runnable).


function bench_cost_model()

% Dispatch hints would print inside the timed closures and distort
% the measurements; silence them for the run (session-scoped).
mptDefaults('showHints', false);
internal.maybeShowTruncationNotice('suppress');
outPath = fullfile(fileparts(mfilename('fullpath')), ...
    'bench_cost_model_matlab.csv');
fid = fopen(outPath, 'w');
fprintf(fid, ['language,part,r,K,N,regime,chosen_off,chosen_on,' ...
              'pw_off,orbit_off,pw_on,orbit_on,ms\n']);

%% ---- Part 1: selector audit ----
fprintf('== selector skip-flag audit (abs, A=1, N_x=N_y) ==\n');
flips = zeros(0, 3);
flipRoutes = cell(0, 2);
for r = [2 3]
    for N = [4 16]
        for K = [4 6 8 12 16 24 40]
            [cOff, pwOff, orbOff] = internal.selectMaInnerProductMethod( ...
                r, K, 1, N, N, false, false, false, 0.0, 'auto', ...
                false, false, [], K, {}, [], ...
                false, false, false, false);
            [cOn, pwOn, orbOn] = internal.selectMaInnerProductMethod( ...
                r, K, 1, N, N, false, false, false, 0.0, 'auto', ...
                false, false, [], K, {}, [], ...
                true, true, true, true);
            isFlip = ~strcmp(cOff, cOn);
            if isFlip
                flips(end+1, :) = [r, K, N];              %#ok<SAGROW>
                flipRoutes(end+1, :) = {cOff, cOn};       %#ok<SAGROW>
            end
            fprintf(fid, ['matlab,selector,%d,%d,%d,,%s,%s,' ...
                          '%.4g,%.4g,%.4g,%.4g,\n'], ...
                r, K, N, cOff, cOn, pwOff, orbOff, pwOn, orbOn);
            if isFlip
                mark = '  <-- flip';
            else
                mark = '';
            end
            fprintf('  r=%d K=%3d N=%3d: off=%-6s on=%-6s%s\n', ...
                r, K, N, cOff, cOn, mark);
        end
    end
end

%% ---- Part 1b: selector audit, asymmetric (the broadcast regime) ----
% A large shared X against a small fixed query (K_y = 4, N_y = 3): the
% shared self matrix dominates the full-triple Bulger price, so
% skipping it (a later broadcast pair) is where the routing genuinely
% moves.
fprintf('\n== selector audit, asymmetric (K_y=4, N_y=3) ==\n');
for r = [2 3]
    for Nx = [8 32]
        for Kx = [6 8 12 16 24 40 64]
            [cOff, pwOff, orbOff] = internal.selectMaInnerProductMethod( ...
                r, Kx, 1, Nx, 3, false, false, false, 0.0, 'auto', ...
                false, false, [], 4, {}, [], ...
                false, false, false, false);
            [cOn, pwOn, orbOn] = internal.selectMaInnerProductMethod( ...
                r, Kx, 1, Nx, 3, false, false, false, 0.0, 'auto', ...
                false, false, [], 4, {}, [], ...
                true, true, true, true);
            isFlip = ~strcmp(cOff, cOn);
            if isFlip
                flips(end+1, :) = [r, Kx, Nx];            %#ok<SAGROW>
                flipRoutes(end+1, :) = {cOff, cOn};       %#ok<SAGROW>
            end
            fprintf(fid, ['matlab,selector_asym,%d,%d,%d,Ky4_Ny3,%s,%s,' ...
                          '%.4g,%.4g,%.4g,%.4g,\n'], ...
                r, Kx, Nx, cOff, cOn, pwOff, orbOff, pwOn, orbOn);
            if isFlip
                mark = '  <-- flip';
            else
                mark = '';
            end
            fprintf('  r=%d Kx=%3d Nx=%3d: off=%-6s on=%-6s%s\n', ...
                r, Kx, Nx, cOff, cOn, mark);
        end
    end
end

%% ---- Part 2: Möbius per-matrix timings ----
fprintf('\n== Möbius per-matrix timings (abs, A=1, K=8) ==\n');
for r = [2 3]
    for N = [20 60]
        dX = iBuildAbs(8, N, r, 0);
        dY = iBuildAbs(8, N, r, 137);

        fnCosFresh = @() cosSimExpTens(dX, dY, ...
            'method', 'mobius', 'verbose', false);
        fnOsdFresh = @() cosSimExpTens(dX, dY, ...
            'method', 'mobius', 'normalize', 'oneSidedDenom', ...
            'verbose', false);
        % Warm regime: seed both self terms once via the cache-carrying
        % outputs, then time calls on the seeded structs — the timed
        % closure computes the cross matrix only (matching the Python
        % bench, whose memo persists on the objects across calls).
        [~, dXw, dYw] = cosSimExpTens(dX, dY, 'method', 'mobius', ...
            'verbose', false);
        fnCosWarm = @() cosSimExpTens(dXw, dYw, 'method', 'mobius', ...
            'verbose', false);

        t3 = internal.timeRepeated(fnCosFresh) * 1000;
        t2 = internal.timeRepeated(fnOsdFresh) * 1000;
        t1 = internal.timeRepeated(fnCosWarm) * 1000;
        fprintf(fid, 'matlab,mobius_matrices,%d,8,%d,cos_fresh_3mat,,,,,,,%.5f\n', r, N, t3);
        fprintf(fid, 'matlab,mobius_matrices,%d,8,%d,osd_fresh_2mat,,,,,,,%.5f\n', r, N, t2);
        fprintf(fid, 'matlab,mobius_matrices,%d,8,%d,cos_warm_1mat,,,,,,,%.5f\n', r, N, t1);
        fprintf(['  r=%d N=%3d: 3mat %.3f  2mat %.3f  1mat %.3f ms' ...
                 '  ->  xy~%.3f  yy~%.3f  xx~%.3f\n'], ...
            r, N, t3, t2, t1, t1, max(t2 - t1, 0), max(t3 - t2, 0));
    end
end

%% ---- Part 3: flip-cell forced-route timings (warm selves) ----
fprintf('\n== flip-cell forced-route timings (warm selves) ==\n');
nCheck = min(size(flips, 1), 4);
for fi = 1:nCheck
    r = flips(fi, 1); K = flips(fi, 2); N = flips(fi, 3);
    % Flip cells arise on the asymmetric grid: shared X (K, N) against
    % the small fixed query (mirrors bench_cost_model.py).
    dX = iBuildAbs(K, N, r, 0);
    dY = iBuildAbs(4, 3, r, 137);
    tForced = struct();
    for methC = {'bulger', 'mobius'}
        meth = methC{1};
        % Seed both self terms under this route once, outside the timed
        % closure (see the part-2 warm-regime note).
        [~, dXw, dYw] = cosSimExpTens(dX, dY, 'method', meth, ...
            'verbose', false);
        fn = @() cosSimExpTens(dXw, dYw, 'method', meth, ...
            'verbose', false);
        t = internal.timeRepeated(fn) * 1000;
        tForced.(meth) = t;
        fprintf(fid, 'matlab,flip_check,%d,%d,%d,forced_%s_warm,%s,%s,,,,,%.5f\n', ...
            r, K, N, meth, flipRoutes{fi, 1}, flipRoutes{fi, 2}, t);
    end
    if tForced.bulger <= tForced.mobius
        faster = 'bulger';
    else
        faster = 'mobius';
    end
    if strcmp(faster, flipRoutes{fi, 2})
        verdict = 'OK';
    else
        verdict = 'MISPICK';
    end
    fprintf(['  r=%d K=%d N=%d: off->%s, on->%s; warm bulger %.3f ms, ' ...
             'mobius %.3f ms; faster=%s [%s]\n'], ...
        r, K, N, flipRoutes{fi, 1}, flipRoutes{fi, 2}, ...
        tForced.bulger, tForced.mobius, faster, verdict);
end

fclose(fid);
fprintf('\nwrote %s\n', outPath);

end


function d = iBuildAbs(K, N, r, seedShift)
%IBUILDABS  Deterministic absolute density (mirrors bench_cost_model.py).
    j = (1:(K * N)) + seedShift;
    vals = reshape(2000 * mod(11 * j.^2 + 5 * j, 397) / 397, K, N);
    d = buildExpTens({vals}, [], 30.0, r, false, false, 0, ...
        'verbose', false);
end

